diff --git a/src/openvpn/init.c b/src/openvpn/init.c
index 1476737e..f396d1ec 100644
--- a/src/openvpn/init.c
+++ b/src/openvpn/init.c
@@ -1707,7 +1707,7 @@ do_route(const struct options *options,
          openvpn_net_ctx_t *ctx)
 {
     bool ret = true;
-    if (!options->route_noexec && ( route_list || route_ipv6_list ) )
+    if (!options->route_noexec && !tt->is_pipe && ( route_list || route_ipv6_list ) )
     {
         ret = add_routes(route_list, route_ipv6_list, tt, ROUTE_OPTION_FLAGS(options),
                          es, ctx);
@@ -1842,12 +1842,12 @@ do_open_tun(struct context *c, int *error_flags)
 
         /* parse and resolve the route option list */
         ASSERT(c->c2.link_socket);
-        if (c->options.routes && c->c1.route_list)
+        if (c->options.routes && c->c1.route_list && !c->c1.tuntap->is_pipe)
         {
             do_init_route_list(&c->options, c->c1.route_list,
                                &c->c2.link_socket->info, c->c2.es, &c->net_ctx);
         }
-        if (c->options.routes_ipv6 && c->c1.route_ipv6_list)
+        if (c->options.routes_ipv6 && c->c1.route_ipv6_list && !c->c1.tuntap->is_pipe)
         {
             do_init_route_ipv6_list(&c->options, c->c1.route_ipv6_list,
                                     &c->c2.link_socket->info, c->c2.es,
@@ -1855,7 +1855,9 @@ do_open_tun(struct context *c, int *error_flags)
         }
 
         /* do ifconfig */
+        c->c1.tuntap->mtu = c->c2.frame.tun_mtu;
         if (!c->options.ifconfig_noexec
+            && !c->c1.tuntap->is_pipe
             && ifconfig_order() == IFCONFIG_BEFORE_TUN_OPEN)
         {
             /* guess actual tun/tap unit number that will be returned
@@ -1864,12 +1866,12 @@ do_open_tun(struct context *c, int *error_flags)
                                                  c->options.dev_type,
                                                  c->options.dev_node,
                                                  &gc);
-            do_ifconfig(c->c1.tuntap, guess, c->c2.frame.tun_mtu, c->c2.es,
-                        &c->net_ctx);
+            do_ifconfig(c->c1.tuntap, guess, c->c2.es, &c->net_ctx);
         }
 
         /* possibly add routes */
-        if (route_order() == ROUTE_BEFORE_TUN)
+        if (route_order() == ROUTE_BEFORE_TUN
+            && !c->c1.tuntap->is_pipe)
         {
             /* Ignore route_delay, would cause ROUTE_BEFORE_TUN to be ignored */
             bool status = do_route(&c->options, c->c1.route_list, c->c1.route_ipv6_list,
@@ -1890,7 +1892,7 @@ do_open_tun(struct context *c, int *error_flags)
                  c->c1.tuntap, &c->net_ctx);
 
         /* set the hardware address */
-        if (c->options.lladdr)
+        if (c->options.lladdr && !c->c1.tuntap->is_pipe)
         {
             set_lladdr(&c->net_ctx, c->c1.tuntap->actual_name, c->options.lladdr,
                        c->c2.es);
@@ -1898,10 +1900,11 @@ do_open_tun(struct context *c, int *error_flags)
 
         /* do ifconfig */
         if (!c->options.ifconfig_noexec
+            && !c->c1.tuntap->is_pipe
             && ifconfig_order() == IFCONFIG_AFTER_TUN_OPEN)
         {
             do_ifconfig(c->c1.tuntap, c->c1.tuntap->actual_name,
-                        c->c2.frame.tun_mtu, c->c2.es, &c->net_ctx);
+                        c->c2.es, &c->net_ctx);
         }
 
         /* run the up script */
diff --git a/src/openvpn/run_command.c b/src/openvpn/run_command.c
index 292e81f7..ee078437 100644
--- a/src/openvpn/run_command.c
+++ b/src/openvpn/run_command.c
@@ -148,6 +148,13 @@ openvpn_execve(const struct argv *a, const struct env_set *es, const unsigned in
             pid = fork();
             if (pid == (pid_t)0) /* child side */
             {
+                if (flags & S_SETPGRP)
+                {
+                    if (setpgid(0, getpid()) == -1)
+                    {
+                        msg(M_WARN | M_ERRNO, "openvpn_execve: setpgid failed");
+                    }
+                }
                 execve(cmd, argv, envp);
                 exit(OPENVPN_EXECVE_FAILURE);
             }
@@ -157,7 +164,11 @@ openvpn_execve(const struct argv *a, const struct env_set *es, const unsigned in
             }
             else /* parent side */
             {
-                if (waitpid(pid, &ret, 0) != pid)
+                if (flags & S_NOWAIT)
+                {
+                    ret = (int) pid;
+                }
+                else if (waitpid(pid, &ret, 0) != pid)
                 {
                     ret = OPENVPN_EXECVE_ERROR;
                 }
@@ -204,6 +215,11 @@ openvpn_execve_check(const struct argv *a, const struct env_set *es, const unsig
             goto done;
         }
     }
+    else if (flags & S_NOWAIT)
+    {
+        ret = stat;
+        goto done;
+    }
     else if (platform_system_ok(stat))
     {
         ret = true;
diff --git a/src/openvpn/run_command.h b/src/openvpn/run_command.h
index 52ab1155..91de5377 100644
--- a/src/openvpn/run_command.h
+++ b/src/openvpn/run_command.h
@@ -47,6 +47,8 @@ void script_security_set(int level);
 /** Instead of returning 1/0 for success/fail,
  * return exit code when between 0 and 255 and -1 otherwise */
 #define S_EXITCODE  (1<<2)
+#define S_NOWAIT    (1<<3)
+#define S_SETPGRP   (1<<4)
 
 /* wrapper around the execve() call */
 int openvpn_popen(const struct argv *a,  const struct env_set *es);
diff --git a/src/openvpn/tun.c b/src/openvpn/tun.c
index ad7b6e9c..2b1338f6 100644
--- a/src/openvpn/tun.c
+++ b/src/openvpn/tun.c
@@ -459,6 +459,10 @@ dev_type_enum(const char *dev, const char *dev_type)
     {
         return DEV_TYPE_NULL;
     }
+    else if (dev && *dev == '|')
+    {
+        return DEV_TYPE_TUN;
+    }
     else
     {
         return DEV_TYPE_UNDEF;
@@ -841,6 +845,11 @@ init_tun(const char *dev,        /* --dev option */
     tt->type = dev_type_enum(dev, dev_type);
     tt->topology = topology;
 
+    if (dev && *dev == '|')
+    {
+        tt->is_pipe = true;
+    }
+
     if (ifconfig_local_parm && ifconfig_remote_netmask_parm)
     {
         bool tun = false;
@@ -1661,9 +1670,10 @@ do_ifconfig_ipv4(struct tuntap *tt, const char *ifname, int tun_mtu,
 
 /* execute the ifconfig command through the shell */
 void
-do_ifconfig(struct tuntap *tt, const char *ifname, int tun_mtu,
+do_ifconfig(struct tuntap *tt, const char *ifname,
             const struct env_set *es, openvpn_net_ctx_t *ctx)
 {
+    int tun_mtu = tt->mtu;
     msg(D_LOW, "do_ifconfig, ipv4=%d, ipv6=%d", tt->did_ifconfig_setup,
         tt->did_ifconfig_ipv6_setup);
 
@@ -1799,6 +1809,64 @@ open_null(struct tuntap *tt)
     tt->actual_name = string_alloc("null", NULL);
 }
 
+static void
+set_vpnc_vars(struct env_set *es, struct tuntap *tt)
+{
+    struct gc_arena gc = gc_new();
+
+    setenv_str(es, "INTERNAL_IP4_ADDRESS", (char *) print_in_addr_t(tt->local, 0, &gc));
+    setenv_int(es, "INTERNAL_IP4_MTU", tt->mtu);
+
+    if (tt->did_ifconfig_ipv6_setup)
+    {
+        const char *ifconfig_ipv6_local = print_in6_addr(tt->local_ipv6, 0, &gc);
+        struct buffer out6 = alloc_buf_gc(64, &gc);
+
+        buf_printf(&out6, "%s/%d", ifconfig_ipv6_local, tt->netbits_ipv6);
+        setenv_str(es, "INTERNAL_IP6_NETMASK", (char *) buf_bptr(&out6));
+    }
+
+    gc_free(&gc);
+}
+
+static void
+open_pipe(const char *dev, struct tuntap *tt)
+{
+    struct argv argv;
+    struct env_set *es;
+    int fds[2], pid;
+
+    if (socketpair(AF_UNIX, SOCK_DGRAM, 0, fds) == -1)
+    {
+        msg(M_FATAL | M_ERRNO, "ERROR: socketpair call failed");
+    }
+
+    tt->fd = fds[0];
+    tt->actual_name = string_alloc("pipe", NULL);
+
+    set_nonblock(tt->fd);
+    set_cloexec(tt->fd);
+
+    es = env_set_create(NULL);
+    setenv_int(es, "VPNFD", fds[1]);
+    set_vpnc_vars(es, tt);
+
+    argv = argv_new();
+    /* dev looks like: "|/path/to/program <args...>" */
+    argv_printf(&argv, "/bin/sh -c %s", &dev[1]);
+    pid = openvpn_execve_check(&argv, es, M_ERR | S_SCRIPT | S_NOWAIT | S_SETPGRP,
+                               "ERROR: Unable to execute TUN script");
+    argv_free(&argv);
+    env_set_destroy(es);
+    close(fds[1]);
+
+    if (pid <= 0)
+    {
+        msg(M_FATAL, "ERROR: unable to start subprocess");
+    }
+    tt->pipe_pid = (pid_t) pid;
+}
+
 
 #if defined (TARGET_OPENBSD) || (defined(TARGET_DARWIN) && HAVE_NET_IF_UTUN_H)
 
@@ -1915,7 +1983,11 @@ open_tun_generic(const char *dev, const char *dev_type, const char *dev_node,
     char dynamic_name[256];
     bool dynamic_opened = false;
 
-    if (tt->type == DEV_TYPE_NULL)
+    if (tt->is_pipe)
+    {
+        open_pipe(dev, tt);
+    }
+    else if (tt->type == DEV_TYPE_NULL)
     {
         open_null(tt);
     }
@@ -2077,6 +2149,10 @@ close_tun_generic(struct tuntap *tt)
     }
 
     free(tt->actual_name);
+    if (tt->pipe_pid)
+    {
+        kill(-tt->pipe_pid, SIGHUP);
+    }
     clear_tuntap(tt);
 }
 #endif /* !_WIN32 */
@@ -2186,11 +2262,15 @@ open_tun(const char *dev, const char *dev_type, const char *dev_node, struct tun
 {
     struct ifreq ifr;
 
-    /*
-     * We handle --dev null specially, we do not open /dev/null for this.
-     */
-    if (tt->type == DEV_TYPE_NULL)
+    if (tt->is_pipe)
     {
+        open_pipe(dev, tt);
+    }
+    else if (tt->type == DEV_TYPE_NULL)
+    {
+        /*
+         * We handle --dev null specially, we do not open /dev/null for this.
+         */
         open_null(tt);
     }
     else if (tun_dco_enabled(tt))
@@ -2372,6 +2452,13 @@ close_tun(struct tuntap *tt, openvpn_net_ctx_t *ctx)
 {
     ASSERT(tt);
 
+    if (tt->pipe_pid)
+    {
+        close_tun_generic(tt);
+        free(tt);
+        return;
+    }
+
 #if defined(TARGET_LINUX) || defined(TARGET_FREEBSD)
     if (tun_dco_enabled(tt))
     {
@@ -2419,6 +2506,12 @@ open_tun(const char *dev, const char *dev_type, const char *dev_node, struct tun
      */
     CLEAR(ifr);
 
+    if (tt->is_pipe)
+    {
+        open_pipe(dev, tt);
+        return;
+    }
+
     if (tt->type == DEV_TYPE_NULL)
     {
         open_null(tt);
@@ -3449,6 +3542,13 @@ close_tun(struct tuntap *tt, openvpn_net_ctx_t *ctx)
 {
     ASSERT(tt);
 
+    if (tt->pipe_pid)
+    {
+        close_tun_generic(tt);
+        free(tt);
+        return;
+    }
+
     struct gc_arena gc = gc_new();
     struct argv argv = argv_new();
 
@@ -3505,6 +3605,12 @@ open_tun(const char *dev, const char *dev_type, const char *dev_node, struct tun
     char dynamic_name[20];
     const char *p;
 
+    if (tt->is_pipe)
+    {
+        open_pipe(dev, tt);
+        return;
+    }
+
     if (tt->type == DEV_TYPE_NULL)
     {
         open_null(tt);
@@ -3598,6 +3704,13 @@ close_tun(struct tuntap *tt, openvpn_net_ctx_t *ctx)
 {
     ASSERT(tt);
 
+    if (tt->pipe_pid)
+    {
+        close_tun_generic(tt);
+        free(tt);
+        return;
+    }
+
     struct argv argv = argv_new();
     struct env_set *es = env_set_create(NULL);
 
@@ -6852,7 +6965,12 @@ open_tun(const char *dev, const char *dev_type, const char *dev_node, struct tun
 
     msg( M_INFO, "open_tun");
 
-    if (tt->type == DEV_TYPE_NULL)
+    if (tt->is_pipe)
+    {
+        open_pipe(dev, tt);
+        return;
+    }
+    else if (tt->type == DEV_TYPE_NULL)
     {
         open_null(tt);
         return;
@@ -7013,6 +7131,13 @@ close_tun(struct tuntap *tt, openvpn_net_ctx_t *ctx)
 {
     ASSERT(tt);
 
+    if (tt->pipe_pid)
+    {
+        close_tun_generic(tt);
+        free(tt);
+        return;
+    }
+
     struct gc_arena gc = gc_new();
 
     if (tt->did_ifconfig_ipv6_setup)
diff --git a/src/openvpn/tun.h b/src/openvpn/tun.h
index 91dbeefe..81de4b1d 100644
--- a/src/openvpn/tun.h
+++ b/src/openvpn/tun.h
@@ -177,6 +177,8 @@ struct tuntap
 #define TUNNEL_TOPOLOGY(tt) ((tt) ? ((tt)->topology) : TOP_UNDEF)
     int topology; /* one of the TOP_x values */
 
+    bool is_pipe;
+    pid_t pipe_pid;
     bool did_ifconfig_setup;
     bool did_ifconfig_ipv6_setup;
 
@@ -189,6 +191,7 @@ struct tuntap
     /* ifconfig parameters */
     in_addr_t local;
     in_addr_t remote_netmask;
+    int mtu;
 
     struct in6_addr local_ipv6;
     struct in6_addr remote_ipv6;
@@ -315,11 +318,10 @@ void do_ifconfig_setenv(const struct tuntap *tt,
  *
  * @param tt        the tuntap interface context
  * @param ifname    the human readable interface name
- * @param mtu       the MTU value to set the interface to
  * @param es        the environment to be used when executing the commands
  * @param ctx       the networking API opaque context
  */
-void do_ifconfig(struct tuntap *tt, const char *ifname, int tun_mtu,
+void do_ifconfig(struct tuntap *tt, const char *ifname,
                  const struct env_set *es, openvpn_net_ctx_t *ctx);
 
 /**
